### Завдання 2

Проведіть крос-валідацію для моделі, побудованої на основі датафрейму  petrol_consumption. Обґрунтуйте обраний метод крос-валідації.

In [1]:
import pandas as pd
import numpy as np
import sklearn
import openpyxl
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import KFold, StratifiedKFold, cross_val_score
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, mean_squared_error

In [30]:
df=pd.read_csv('https://raw.githubusercontent.com/Startup-acdm/PDS1/3810a4a8a7a352810d551cb3da52310a7dab1db8/Lesson%2035/petrol_consumption.csv?token=GHSAT0AAAAAAB5FDR3Z5NKYCBQQBDNDL4Y4Y7WOFEQ')
df.head(5)

Petrol_tax  Average_income  Paved_Highways  Population_Driver_licence(%)  \
0         9.0            3571            1976                         0.525   
1         9.0            4092            1250                         0.572   
2         9.0            3865            1586                         0.580   
3         7.5            4870            2351                         0.529   
4         8.0            4399             431                         0.544   

   Petrol_Consumption  
0                 541  
1                 524  
2                 561  
3                 414  
4                 410

In [31]:
df.describe()

Petrol_tax  Average_income  Paved_Highways  \
count   48.000000       48.000000       48.000000   
mean     7.668333     4241.833333     5565.416667   
std      0.950770      573.623768     3491.507166   
min      5.000000     3063.000000      431.000000   
25%      7.000000     3739.000000     3110.250000   
50%      7.500000     4298.000000     4735.500000   
75%      8.125000     4578.750000     7156.000000   
max     10.000000     5342.000000    17782.000000   

       Population_Driver_licence(%)  Petrol_Consumption  
count                     48.000000           48.000000  
mean                       0.570333          576.770833  
std                        0.055470          111.885816  
min                        0.451000          344.000000  
25%                        0.529750          509.500000  
50%                        0.564500          568.500000  
75%                        0.595250          632.750000  
max                        0.724000          968.000000

In [6]:
X=np.array(df[['Petrol_tax','Average_income','Paved_Highways','Population_Driver_licence(%)']])
y=np.array(df[['Petrol_Consumption']])

In [10]:
cor=df.corr()
cor

Petrol_tax  Average_income  Paved_Highways  \
Petrol_tax                      1.000000        0.012665       -0.522130   
Average_income                  0.012665        1.000000        0.050163   
Paved_Highways                 -0.522130        0.050163        1.000000   
Population_Driver_licence(%)   -0.288037        0.157070       -0.064129   
Petrol_Consumption             -0.451280       -0.244862        0.019042   

                              Population_Driver_licence(%)  Petrol_Consumption  
Petrol_tax                                       -0.288037           -0.451280  
Average_income                                    0.157070           -0.244862  
Paved_Highways                                   -0.064129            0.019042  
Population_Driver_licence(%)                      1.000000            0.698965  
Petrol_Consumption                                0.698965            1.000000

#### Проведемо розрахунки для вибірки з використанням функції train_test_split та побудуємо різні моделі

#### Модель 1 з параметрами  
test_size=0.2  
random_state=None  
shuffle=True  

In [37]:
model = LinearRegression()
X_train, X_test, y_train, y_test = sklearn.model_selection.train_test_split(X, y, test_size=0.2)
model.fit(X_train,y_train)
a=model.intercept_[0]
b=model.coef_[0]
y_pred=model.predict(X_test)
y_pred_all=model.predict(X)
print('Функція регресії має вигляд - Y={:.4f}+{:.4f}*X1+{:.4f}*X2+{:.4f}*X3+{:.4f}*X4'.format(a,b[0],b[1],b[2],b[3]))
print('Достовірність моделі на тренувальних даних: Score= {:.4f}'.format(model.score(X_train, y_train)))
print('Достовірність моделі на тестових даних: Score= {:.4f}'.format(model.score(X_test, y_test)))
print('Достовірність моделі на повній виборці: Score= {:.4f}'.format(model.score(X, y)))
print(f'mae = {mean_absolute_error(y_test,y_pred):4f}' )
print(f'mse_test = {mean_squared_error(y_test,y_pred):4f}, mse_all = {mean_squared_error(y,y_pred_all):4f}')
mrse=mean_squared_error(y_test,y_pred)**0.5
print(f'mrse= {mrse:4f}')

Функція регресії має вигляд - Y=378.8611+-37.5736*X1+-0.0580*X2+-0.0042*X3+1333.8606*X4
Достовірність моделі на тренувальних даних: Score= 0.7293
Достовірність моделі на тестових даних: Score= 0.3606
Достовірність моделі на повній виборці: Score= 0.6726
mae = 67.863546
mse_test = 6191.482912, mse_all = 4013.084714
mrse= 78.685977


#### Модель 2 з параметрами 

test_size=0.2  
random_state=None  
shuffle=False 

In [38]:
X_train, X_test, y_train, y_test = sklearn.model_selection.train_test_split(X, y, test_size=0.2,  shuffle=False)
model.fit(X_train,y_train)
a=model.intercept_[0]
b=model.coef_[0]
y_pred=model.predict(X_test)
y_pred_all=model.predict(X)
print('Функція регресії має вигляд - Y={:.4f}+{:.4f}*X1+{:.4f}*X2+{:.4f}*X3+{:.4f}*X4'.format(a,b[0],b[1],b[2],b[3]))
print('Достовірність моделі на тренувальних даних: Score= {:.4f}'.format(model.score(X_train, y_train)))
print('Достовірність моделі на тестових даних: Score= {:.4f}'.format(model.score(X_test, y_test)))
print('Достовірність моделі на повній виборці: Score= {:.4f}'.format(model.score(X, y)))
print(f'mae = {mean_absolute_error(y_test,y_pred):4f}' )
print(f'mse_test = {mean_squared_error(y_test,y_pred):4f}, mse_all = {mean_squared_error(y,y_pred_all):4f}')
mrse=mean_squared_error(y_test,y_pred)**0.5
print(f'mrse= {mrse:4f}')


Функція регресії має вигляд - Y=218.7732+-12.2979*X1+-0.0729*X2+0.0029*X3+1292.5094*X4
Достовірність моделі на тренувальних даних: Score= 0.7231
Достовірність моделі на тестових даних: Score= 0.2919
Достовірність моделі на повній виборці: Score= 0.6382
mae = 79.582309
mse_test = 11721.964814, mse_all = 4434.381296
mrse= 108.268023


#### Модель 3 з параметрами 

test_size=0.2  
random_state=48    
shuffle=True 

In [39]:
X_train, X_test, y_train, y_test = sklearn.model_selection.train_test_split(X, y, test_size=0.2,  random_state=48)
model.fit(X_train,y_train)
a=model.intercept_[0]
b=model.coef_[0]
y_pred=model.predict(X_test)
y_pred_all=model.predict(X)
print('Функція регресії має вигляд - Y={:.4f}+{:.4f}*X1+{:.4f}*X2+{:.4f}*X3+{:.4f}*X4'.format(a,b[0],b[1],b[2],b[3]))
print('Достовірність моделі на тренувальних даних: Score= {:.4f}'.format(model.score(X_train, y_train)))
print('Достовірність моделі на тестових даних: Score= {:.4f}'.format(model.score(X_test, y_test)))
print('Достовірність моделі на повній виборці: Score= {:.4f}'.format(model.score(X, y)))
print(f'mae = {mean_absolute_error(y_test,y_pred):4f}' )
print(f'mse_test = {mean_squared_error(y_test,y_pred):4f}, mse_all = {mean_squared_error(y,y_pred_all):4f}')
mrse=mean_squared_error(y_test,y_pred)**0.5
print(f'mrse= {mrse:4f}')

Функція регресії має вигляд - Y=344.5859+-35.3997*X1+-0.0587*X2+-0.0037*X3+1370.9393*X4
Достовірність моделі на тренувальних даних: Score= 0.6896
Достовірність моделі на тестових даних: Score= 0.4282
Достовірність моделі на повній виборці: Score= 0.6690
mae = 58.481805
mse_test = 3983.845194, mse_all = 4057.778839
mrse= 63.117709


### Проведемо розрахунки з використанням k-Fold кросс-валідації та розіб'ємо виборку на 5 частин

In [49]:
kfold = KFold(n_splits=5)
score = cross_val_score(model, X, y, cv=kfold)
score

array([-0.22573437,  0.62535306, -0.08289785, -0.16796305,  0.30273973])

In [50]:
for (train_index, test_index) in kfold.split(X):
    X_train=X[train_index]
    y_train=y[train_index]
    X_test=X[test_index]
    y_test=y[test_index]
   
    model.fit(X_train,y_train) 
    a=model.intercept_[0]
    b=model.coef_[0]
    print('Функція регресії має вигляд - Y={:.4f}+{:.4f}*X1+{:.4f}*X2+{:.4f}*X3+{:.4f}*X4'.format(a,b[0],b[1],b[2],b[3]))
    print('Достовірність моделі на тренувальних даних: Score= {:.4f}'.format(model.score(X_train, y_train)))
    print('Достовірність моделі на тестових даних: Score= {:.4f}'.format(model.score(X_test, y_test)))
    print('Достовірність моделі на повній виборці: Score= {:.4f}'.format(model.score(X, y)))
    y_pred=model.predict(X_test)
    y_pred_all=model.predict(X)
    print(f'mae = {mean_absolute_error(y_test,y_pred):4f}' )
    print(f'mse_test = {mean_squared_error(y_test,y_pred):4f}, mse_all = {mean_squared_error(y,y_pred_all):4f}')
    mrse=mean_squared_error(y_test,y_pred)**0.5
    print(f'mrse= {mrse:4f}\n')

Функція регресії має вигляд - Y=495.0632+-42.4414*X1+-0.0533*X2+-0.0063*X3+1189.0250*X4
Достовірність моделі на тренувальних даних: Score= 0.6160
Достовірність моделі на тестових даних: Score= -0.2257
Достовірність моделі на повній виборці: Score= 0.6566
mae = 51.002943
mse_test = 4879.403398, mse_all = 4209.463034
mrse= 69.852726

Функція регресії має вигляд - Y=354.9385+-32.5960*X1+-0.0665*X2+-0.0024*X3+1338.5491*X4
Достовірність моделі на тренувальних даних: Score= 0.6801
Достовірність моделі на тестових даних: Score= 0.6254
Достовірність моделі на повній виборці: Score= 0.6771
mae = 57.571201
mse_test = 4390.716011, mse_all = 3957.933213
mrse= 66.262478

Функція регресії має вигляд - Y=380.1415+-46.0102*X1+-0.0597*X2+-0.0037*X3+1433.3849*X4
Достовірність моделі на тренувальних даних: Score= 0.7045
Достовірність моделі на тестових даних: Score= -0.0829
Достовірність моделі на повній виборці: Score= 0.6627
mae = 51.957592
mse_test = 3639.359791, mse_all = 4133.896954
mrse= 60.327107


### Побудуємо модель стратифікованої К-кратної крос-валідації із застосуванням классу StratifiedKFold (n_splits=2)

In [51]:
skf = StratifiedKFold(n_splits=2)
for (train_index, test_index) in skf.split(X,y):
    X_train=X[train_index]
    y_train=y[train_index]
    X_test=X[test_index]
    y_test=y[test_index]
   
    model.fit(X_train,y_train) 
    a=model.intercept_[0]
    b=model.coef_[0]
    print('Функція регресії має вигляд - Y={:.4f}+{:.4f}*X1+{:.4f}*X2+{:.4f}*X3+{:.4f}*X4'.format(a,b[0],b[1],b[2],b[3]))
    print('Достовірність моделі на тренувальних даних: Score= {:.4f}'.format(model.score(X_train, y_train)))
    print('Достовірність моделі на тестових даних: Score= {:.4f}'.format(model.score(X_test, y_test)))
    print('Достовірність моделі на повній виборці: Score= {:.4f}'.format(model.score(X, y)))
    y_pred=model.predict(X_test)
    y_pred_all=model.predict(X)
    print(f'mae = {mean_absolute_error(y_test,y_pred):4f}' )
    print(f'mse_test = {mean_squared_error(y_test,y_pred):4f}, mse_all = {mean_squared_error(y,y_pred_all):4f}')
    mrse=mean_squared_error(y_test,y_pred)**0.5
    print(f'mrse= {mrse:4f}\n')

Функція регресії має вигляд - Y=33.6870+-22.9554*X1+-0.0382*X2+0.0013*X3+1497.9061*X4
Достовірність моделі на тренувальних даних: Score= 0.8076
Достовірність моделі на тестових даних: Score= 0.4225
Достовірність моделі на повній виборці: Score= 0.6077
mae = 65.308619
mse_test = 7368.980181, mse_all = 4808.985977
mrse= 85.842764

Функція регресії має вигляд - Y=473.0360+-42.3506*X1+-0.1032*X2+-0.0092*X3+1635.3265*X4
Достовірність моделі на тренувальних даних: Score= 0.6752
Достовірність моделі на тестових даних: Score= 0.4384
Достовірність моделі на повній виборці: Score= 0.5631
mae = 66.606983
mse_test = 6566.486980, mse_all = 5355.365759
mrse= 81.033863



C:\Python\Python310\lib\site-packages\sklearn\model_selection\_split.py:700: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=2.
  warnings.warn(


### Висновки

#### З усих застосованих методів найкращі показники отримали при  використанні п'ятикратної крос-валідації. Найбільш вдалою є модель, отримана на другому етапі розбивки даних.

Функція регресії має вигляд - Y=354.9385+-32.5960*X1+-0.0665*X2+-0.0024*X3+1338.5491*X4  
Достовірність моделі на тренувальних даних: Score= 0.6801  
Достовірність моделі на тестових даних: Score= 0.6254  
Достовірність моделі на повній виборці: Score= 0.6771  
mae = 57.571201  
mse_test = 4390.716011, mse_all = 3957.933213  
mrse= 66.262478  

####  Достовірність моделі дуже низька, тому модель не може бути використана у практичних розрахунках.
